### PR TITLE
remote target and transmission protocol

### PR DESCRIPTION
**Pull request by ettec**
ks-119

### CODE REVIEW

File: core/capabilities/remote/target/caller.go
Comment (krehermann) [inline core/capabilities/remote/target/caller.go]: why isn't this implemented?

File: core/capabilities/remote/target/caller.go
Comment (krehermann) [inline core/capabilities/remote/target/caller.go]: what is the significance of this interface? 
- why bother with an interface rather than a struct if it's private?
- what does it mean for a 'request' to have an 'OnMessage'? isn't a request a kind of message?

File: core/capabilities/remote/target/caller.go
Comment (krehermann) [inline core/capabilities/remote/target/caller.go]: can you expand this comment with godoc links to the capability API and network message definitions as well an explanation about why we need the shim?

File: core/capabilities/remote/target/receiver.go
Comment (krehermann) [inline core/capabilities/remote/target/receiver.go]: why is this asymmetric WRT the callerRequest?

File: core/capabilities/remote/target/request/caller_request.go
Comment (krehermann) [inline core/capabilities/remote/target/request/caller_request.go]: this is odd. what is an Error_OK? does that mean no error? it seems like an oxymoron.

File: core/capabilities/remote/target/request/caller_request.go
Comment (krehermann) [inline core/capabilities/remote/target/request/caller_request.go]: nit/ you can simplify this and the next branch slightly ~ 
```
alreadySent, exists := c.responseReceived[sender]
if !exists {}
if alreadySent{}
```

File: core/capabilities/remote/target/request/caller_request.go
Comment (krehermann) [inline core/capabilities/remote/target/request/caller_request.go]: maybe better to accept an error than a string for type continuity, especially if the receiver of the capability response needs to handle specific kinds of errors

File: core/capabilities/transmission/transmission.go
Comment (krehermann) [inline core/capabilities/transmission/transmission.go:54]: why is the sharedSecret passed in? presumably it's quite important that it's not some other 16 byte thing, but i can't tell why. can you add a comment please

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: good point, forgot to follow up on that.  @cedric-cordenier I believe this should be a no-op in this case, can you confirm?

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: yeah, no need to be an interface, left over from some refactor I guess, removed.

File: core/capabilities/remote/target/receiver.go
Comment (ettec) [reply core/capabilities/remote/target/receiver.go]: removed

File: core/capabilities/remote/target/request/caller_request.go
Comment (ettec) [reply core/capabilities/remote/target/request/caller_request.go]: I'll defer to @bolekk on that one, guess it could be StatusCode.

Comment (bolekk) [reply core/capabilities/remote/target/request/caller_request.go]: yeah, it should be called status code, I can rename in a separate PR

File: core/capabilities/remote/target/caller.go
Comment (bolekk) [inline core/capabilities/remote/target/caller.go]: maybe add comments that these two are no-ops on purpose

File: core/capabilities/remote/target/receiver.go
Comment (bolekk) [inline core/capabilities/remote/target/receiver.go]: I'm still not a fan of this. I'm worried that it'll make debugging harder. IMO a more natural logic will be to cluster messages based on WorkflowExecutionID and then pick F+1 identical payloads once we have them. Your logic achieves the same outcome so I'm not going to block it.

File: core/capabilities/remote/target/request/receiver_request.go
Comment (bolekk) [inline core/capabilities/remote/target/request/receiver_request.go]: Could you maybe add a comment in OnMessage() explaining that once an execution starts, all subsequent incoming messages will be blocked until it finishes? It wasn't obvious to me at first and I thought a slow underlying execution can cause problems, but it seems to be fine.

**Review by bolekk — DISMISSED** [review]

File: core/capabilities/remote/target/request/receiver_request.go
Comment (ettec) [reply core/capabilities/remote/target/request/receiver_request.go]: Its intentional that a receiverRequest instance is single threaded, and the receiver handles sending each message on a goroutine to ensure the dispatcher thread is not blocked.  I'll add some doc to this effect.

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: done

File: core/capabilities/remote/target/caller.go
Comment (cedric-cordenier) [inline core/capabilities/remote/target/caller.go]: IMO caller isn't a great name :( It makes it unclear about whether this is the object responsible for being the shim (i.e. on the local node) or for being the "bridge" (the object that propagates network calls to the remote capability.

File: core/capabilities/remote/target/caller.go
Comment (cedric-cordenier) [inline core/capabilities/remote/target/caller.go]: Why did you choose to put the Request objects in separate packages? In my mind organising all of the objects that talk to one another into the same package feels like it makes more sense (so caller + caller request; receiver + its request).

File: core/capabilities/remote/target/caller.go
Comment (cedric-cordenier) [reply core/capabilities/remote/target/caller.go]: Aside: it would be really great if we could have some clear naming conventions for these objects across all capability types

Eg. for triggers we have a subscriber (on the local node) and a publisher (on the remote); here we have a Caller (on the local node) and a receiver (on the remote)

What's slightly confusing about caller + receiver is that actually we communicate both ways (i.e. the receiver is also a caller when it's transmitting results back to the local node)

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: to prevent accidental call of non-threadsafe methods, just caution really

File: core/capabilities/remote/target/caller.go
Comment (cedric-cordenier) [inline core/capabilities/remote/target/caller.go]: Does this need to be public? AFAICS it's only called by the worker goroutine

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: don't disagree, lacked the imagination to come up with a better name that didn;t clash with other names already taken. with package name its  remote.target.caller.   Open to suggestions though!

File: core/capabilities/remote/target/caller.go
Comment (cedric-cordenier) [reply core/capabilities/remote/target/caller.go]: (Same comment for the Receiver)

Comment (ettec) [reply core/capabilities/remote/target/caller.go]: ta, sorted that

File: core/capabilities/remote/target/caller.go
Comment (ettec) [reply core/capabilities/remote/target/caller.go]: renamed to client/server - an overloaded term, but it fits better I think

**Review by bolekk — APPROVED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**
